### ISSUE TITLE
Free shipping can be applied multiple times

### ISSUE THREAD

**Issue by ikulis** [PrestaShop/PrestaShop#19792]
#### Describe the bug

You can apply the "Free shipping" discount  multiple times on same order. 
Might be related to https://github.com/PrestaShop/PrestaShop/issues/19790 .

#### Expected behavior

Free shipping can be applied only once.

#### Steps to Reproduce

1. Make order where the total price is at least 3x the shipping price (which should be more than 0) 
![image](https://user-images.githubusercontent.com/1073778/84778759-bd80e700-afe3-11ea-8780-1ec1db771bd3.png)
2. go to order in backoffice
3. click "Add a discount"
4. Choose free shipping
5. click "Add a discount"
6. Choose free shipping
7. 2nd discount for shipping was added.
![image](https://user-images.githubusercontent.com/1073778/84778908-f15c0c80-afe3-11ea-89eb-a62bb06c6211.png)

#### Additional information

* PrestaShop version: 1.7.7.x - c7471b7
* PHP version: PHP 7.3.14-1~deb10u1

**Comment by matks** [comment]
I thought I fixed that in https://github.com/PrestaShop/PrestaShop/pull/18321 🤔

**Comment by ikulis** [comment]
@matks maybe the issue is the `free_shipping` is on 2 places: CartRule and OrderCartRule. 
The one in OrderCartRule is 0 and the one in CartRule is 1.

**Comment by florine2623** [comment]
Hi @ikulis ,

I reproduced your issue in PS 1.7.7.0 with your steps.
I’ll add this to the debug roadmap so that it’s fixed. If you have already fixed it on your end or if you think you can do it, please do send us a pull request!

@matks 
In #18321, the issue was in the FO, this one occurs in the BO

Thanks!

**Comment by matks** [comment]
But BO and FO are supposed to use the same logic 😅 ...

Anyway if you reproduce, it means there is something to fix 💪

**Comment by colinegin** [comment]
ping @florine2623 can you confirm this is a regression ? 
I did not reproduce it on 1765, but if you can try on your side, just to be safe ;)

**Comment by florine2623** [comment]
It is not a regression for me, I reproduced it in PS 1.7.6.5 too.
Please check my screenshot :
![Capture d’écran 2020-06-17 à 17 32 01](https://user-images.githubusercontent.com/16019289/84918229-9d732580-b0c0-11ea-9fe9-5a245a3c2272.png)

**Comment by matks** [comment]
Oh, I believe in "Create an order" page it's OK (I just checked), "free shipping" discounts do not stack.

However it might not be wrong in "Edit/view an order" page.

**Comment by matks** [comment]
Yes: confirmed. You can stack "free shipping" discounts in "Edit an order" page 😄

**Comment by matks** [comment]
All right, for the record:
- "add an order" page relies on AddCartRuleToCartHandler whose logic is good which relies on legacy logic to compute the numbers, so when logic was fixed in https://github.com/PrestaShop/PrestaShop/pull/18321 it worked here too
but
- "edit an order" page relies on AddCartRuleToOrderHandler whose logic is bad because it does the computation logic itself so the bugfix https://github.com/PrestaShop/PrestaShop/pull/18321 is not applied

We might look into @zuk3975 work who tried to refactor in AddCartRuleToOrderHandler https://github.com/PrestaShop/PrestaShop/pull/17348 (but could not get it merged) to see if it does the trick

**Comment by matks** [comment]
And here is a Behat scenario that reproduces (and validates) the wrong behavior to be fixed:
https://github.com/PrestaShop/PrestaShop/compare/1.7.7.x...matks:failing-double-freeshipping-scenario?expand=1
```
$ ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s order --tags aaa

  @add-discounts-to-order
  Scenario: Add free shipping discount to order twice, the 2nd discount should not have effects
    Given order "bo_order1" does not have any invoices
    When I add discount to order "bo_order1" with following details:
      | name | free shipping for you |
      | type | free_shipping         |
    And I add discount to order "bo_order1" with following details:
      | name | free shipping for you - again! |
      | type | free_shipping                  |
    Then order "bo_order1" should have following details:
      | total_products           | 23.8  |
      | total_products_wt        | 25.23 |
      | total_shipping           | 7.42  |
      | total_shipping_tax_excl  | 7.0   |
      | total_shipping_tax_incl  | 7.42  |
      | total_discounts_tax_excl | 7.0   |
      | total_discounts_tax_incl | 7.42  |
    Then Order "bo_order1" should have following prices:
      | products  | $23.80 |
      | discounts | $7.00  |
      | shipping  | $7.00  |
      | taxes     | $1.43  |
      | total     | $23.80 |
```
``

**Comment by jolelievre** [comment]
@matks additionally to your behat test, I would another one at the end that adds a product with free shipping (different from adding the discount directly), and checks that the free shipping is still unique

**Comment by colinegin** [comment]
As this is not a regression of 176, this issue is removed from the 177 kanban.
Thanks

**Comment by khouloudbelguith** [comment]
Hi all,

This ticket is a duplicate of #18576
So, I close it.
Thanks!

**Comment by khouloudbelguith** [comment]
Duplicate of #18576
